solidity/ecdsa/contracts/libraries/EcdsaAuthorization.sol

```diff
@@ -15,25 +15,474 @@
     }
 
+    struct AuthorizationDecrease {
+        uint96 decreasingBy; // amount
+        uint64 decreasingAt; // timestamp
+    }
+
+    struct Data {
+        Parameters parameters;
+        mapping(address => address) stakingProviderToOperator;
+        mapping(address => address) operatorToStakingProvider;
+        mapping(address => AuthorizationDecrease) pendingDecreases;
+    }
+
+    event OperatorRegistered(
+        address indexed stakingProvider,
+        address indexed operator
+    );
+
+    event AuthorizationIncreaseRequested(
+        address indexed stakingProvider,
+        address indexed operator,
+        uint96 fromAmount,
+        uint96 toAmount
+    );
+
+    event AuthorizationDecreaseRequested(
+        address indexed stakingProvider,
+        address indexed operator,
+        uint96 fromAmount,
+        uint96 toAmount,
+        uint64 decreasingAt
+    );
+
+    event AuthorizationDecreaseApproved(address indexed stakingProvider);
+
+    event OperatorJoinedSortitionPool(
+        address indexed stakingProvider,
+        address indexed operator
+    );
+
+    event OperatorStatusUpdated(
+        address indexed stakingProvider,
+        address indexed operator
+    );
+
+    /// @notice Sets the minimum authorization for ECDSA application. Without
+    ///         at least the minimum authorization, staking provider is not
+    ///         eligible to join and operate in the network.
     function setMinimumAuthorization(
         Data storage self,
         uint96 _minimumAuthorization
     ) internal {
-        self.minimumAuthorization = _minimumAuthorization;
+        self.parameters.minimumAuthorization = _minimumAuthorization;
     }
 
+    /// @notice Sets the authorization decrease delay. It is the time in seconds
+    ///         that needs to pass between the time authorization decrease is
+    ///         requested and the time the authorization decrease can be
+    ///         approved, no matter the authorization decrease amount.
     function setAuthorizationDecreaseDelay(
         Data storage self,
         uint64 _authorizationDecreaseDelay
     ) internal {
-        self.authorizationDecreaseDelay = _authorizationDecreaseDelay;
+        self
+            .parameters
+            .authorizationDecreaseDelay = _authorizationDecreaseDelay;
+    }
+
+    /// @notice Used by staking provider to set operator address that will
+    ///         operate ECDSA node. The given staking provider can set operator
+    ///         address only one time. The operator address can not be changed
+    ///         and must be unique. Reverts if the operator is already set for
+    ///         the staking provider or if the operator address is already in
+    ///         use. Reverts if there is a pending authorization decrease for
+    ///         the staking provider.
+    function registerOperator(Data storage self, address operator) internal {
+        address stakingProvider = msg.sender;
```

Comment:
We can do it but there are two reasons why I think we shouldn't be doing it.

If `EcdsaAuthorization` is to implement everything related to authorizations in a comprehensive way, then it should expose API that is clear about intentions and provides the flexibility needed but no more. In the case of `joinSortitionPool`, we know that only operator should be able to join the sortition pool and no one else should be able to put an operator into it. Right now, `EcdsaAuthorization` library API is clear about it. Similarly for `registerOperator` - only staking provider should be able to register its operator. Again, API, right now is clear about it.

The second reason is that although we may have enough space in `WalletRegistry` for whatever we need, this might not be the case for the `RandomBeacon` and by not adding these parameters we are possibly making `EcdsaAuthorization` and future `BeaconAuthorization` consistent.

Reply:
Yeah, I see your points, they make sense. I just don't like that we mix where we define the accessibility for functions, sometimes it's modifier in `WalletRegistry` and sometimes it's implemented in a library.

Reply:
I think a good rule is to have a `require` with authorization requirements inside if the library if it is a core logic, for example, only original DKG result submitter being able to approve the result for the first N blocks, and have all `only*` modifiers inside the contract. Unsure if there are any exceptions though and if we are currently implementing it this way everywhere.